jans-cedarling/cedarling/src/tests/mapping_entities.rs

```diff
@@ -295,3 +296,80 @@ fn test_failed_userinfo_token_mapping() {
         "should be error CouldNotFindEntity"
     );
 }
+
+/// Check if we get roles mapping from all tokens.
+/// Because we specify mapping from each token in policy store
+/// We use iss in JWT tokens to enable mapping for trusted issuer in policy store
+#[test]
+fn test_role_many_tokens_mapping() {
+    let raw_config = get_raw_config(POLICY_STORE_RAW_YAML);
+
+    let config = crate::BootstrapConfig::from_raw_config(&raw_config)
+        .expect("raw config should parse without errors");
+
+    let cedarling = Cedarling::new(&config).expect("could be created without error");
+
+    let request = // deserialize `Request` from json
+    Request::deserialize(serde_json::json!(
+        {
+            "access_token": generate_token_using_claims(json!({
+                    "org_id": "some_long_id",
+                    "jti": "some_jti",
+                    "client_id": "some_client_id",
+                    "iss": "https://test-casa.gluu.info",
+                    "aud": "some_aud",
+                    "role": "Guest",
+                  })),
+            "id_token": generate_token_using_claims(json!({
+                    "jti": "some_jti",
+                    "iss": "https://test-casa.gluu.info",
+                    "aud": "some_aud",
+                    "sub": "some_sub",
+                    "role": "User",
+                  })),
+            "userinfo_token":  generate_token_using_claims(json!({
+                    "jti": "some_jti",
+                    "country": "US",
+                    "sub": "some_sub",
+                    "iss": "https://test-casa.gluu.info",
+                    "client_id": "some_client_id",
+                    "role": "Admin",
+                  })),
+            "action": "Jans::Action::\"Update\"",
+            "resource": {
+                "id": "random_id",
+                "type": "Jans::Issue",
+                "org_id": "some_long_id",
+                "country": "US"
+            },
+            "context": {},
+        }
+    ))
+    .expect("Request should be deserialized from json");
+
+    // HashSet of expected roles
+    let mut expected_role_ids: HashSet<String> =
+        HashSet::from_iter(["Guest", "User", "Admin"].into_iter().map(String::from));
+
+    // iterate over roles that created and filter expected roles
+    let roles_left = cedarling
+        .authorize_entities_data(&request)
+        .expect("should get authorize_entities_data without errors")
+        .role_entities
+        .into_iter()
+        .map(|entity| entity.uid().id().escaped())
+        // if role successfully removed from `expected_role_ids` we filter it
+        .filter(|uid| !expected_role_ids.remove(uid.as_str()))
+        .map(|v| v.to_string())
+        .collect::<Vec<_>>();
```

Comment:
maybe it's just not clear to me, but can you explain why if an `access_token` is saying that it's role is a `Guest` and the `Userinfo` token is saying that it's an `Admin`, why is the request still approved? Isn't the information provided for request effectively conflicting?

Reply:
For me, it looks like we should get role only from one token, but Mike said that we should allow get roles from any token.

Even if the information is conflicting the `cedar-policy` will correctly evaluate it.  Because if none of policy is evaluated than result Deny, Permit policy evaluated is Allow and if forbid evaluated result is denied (higher priority).
So it should be OK.

Reply:
and we have no forbid policy

Reply:
> Mike said that we should allow getting roles from any and many tokens.

Can you confirm if it means allowing Cedarling to authorize for multiple roles for through one request or and not because some IDPs won't have all the tokens all the time? Because it doesn't seem right for a User to have, for example, to have a "Guest" and "Admin" role at the same time for one request.

Reply:
> Even if the information is conflicting the cedar-policy will correctly evaluate it.

This seems wrong... why would we allow a User which isn't sure of their identity to be authorized?

> Because if none of policy is evaluated than result Deny

This is right because we should Deny by default.

Reply:
"This is right because we should Deny by default." It is how cedar-policy work, We don't need to do anything special for this

Reply:
`This seems wrong... why would we allow a User which isn't sure of their identity to be authorized?` we write `cedar-policy` rules (policies) to check if we allow or deny.

Reply:
Also, a lot of systems allows users to have many roles simultaneously.

Reply:
> Also, a lot of systems allows users to have many roles simultaneously.

yes, but not from different sources. why not make the "roles" claim for a token a list that contains all the role for that request instead of assigning different roles per token?

Reply:
I think this is a corner case that doesn't exist much in practice. At the end of the day, the Cedarling is open source and people can fork it for their esoteric use cases. One list of roles for a person is enough, e.g. we don't need the roles scoped to a token.